### PR TITLE
PixelPaint: Allow resizing images and layers.

### LINKED ISSUE #11862: PixelPaint: Allow resizing of the image and layers
Currently, you can only *shrink* the image, by using the "Crop to Selection" menu option. Being able to make the image bigger would be very useful. Similarly, we would like to be able to resize a layer.

This would just involve changing the canvas size, without modifying the image data by stretching it. This is not the same as scaling the image/layers, which is already implemented. See below:

image

### PR DESCRIPTION
**Pull request by jpgleeson**
Fixes #11862
This PR adds a None ScalingMode to LibGfx and incorporates that into PixelPaint to allow for resizing of the canvas without scaling the image. The previous behaviour was to always rescale the canvas when resizing an image. This adds a radio button to the resize image dialog, which blits the existing canvas at it's existing size to the top left of the new, resized canvas.

The None ScalingMode also has some code added to handle it in `Gfx::Painter`. This is the same concept, we create a clipped destination rectangle and blit the source image to it, not drawing any of the source that doesn't fit if the destination is smaller, and not stretching the source to the destination rectangle if it is bigger.

### CODE REVIEW

File: Userland/Applications/PixelPaint/Layer.cpp
Comment (AtkinsSJ) [inline Userland/Applications/PixelPaint/Layer.cpp:204]: If we're not scaling the image, we can skip the scaling calculations by calling `painter.blit()` instead.

File: Userland/Applications/PixelPaint/Layer.cpp
Comment (AtkinsSJ) [inline Userland/Applications/PixelPaint/Layer.cpp:217]: `painter.blit()` here too.

File: Userland/Applications/PixelPaint/ResizeImageDialog.cpp
Comment (AtkinsSJ) [inline Userland/Applications/PixelPaint/ResizeImageDialog.cpp:77]: This could be simpler:
```suggestion
        m_rescale_image = is_checked;
```

File: Userland/Applications/PixelPaint/Image.cpp
Comment (AtkinsSJ) [inline Userland/Applications/PixelPaint/Image.cpp:551]: Generally, we prefer to use named enums instead of bools for parameters. Without looking at the function definitions, it's a lot easier to understand something like
```c++
resize(new_size, location(), scaling_mode, ShouldRescale::No);
```
than
```c++
resize(new_size, location(), scaling_mode, false);
```

Thinking about it, you either want to pass a `ScalingMode`, *or* you want to not scale it. Having them be a single parameter would be nice. Maybe a `Gfx::Painter::ScalingMode::None` value, though I'm not sure if that would interfere with other places that use `ScalingMode`.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Welcome to Serenity! :^)

Thanks for tackling this! I do have some suggestions.

Also, if you put `Fixes #11862` on its own line in either the commit message or PR description, GitHub will automatically close that issue when this is merged. Makes things a little easier. :^)

File: Userland/Applications/PixelPaint/Image.cpp
Comment (jpgleeson) [reply Userland/Applications/PixelPaint/Image.cpp]: Good point. I've refactored this PR to use a new `ScalingMode` `None` as suggested e9428e4. This doesn't seem to have caused any issues elsewhere. As a result of that, I've also changed the UI for this to include the Resize Canvas option in the radio selections rather than as a separate checkbox to sit better with what's going on internally and not require special handling if the checkbox was toggled back and forth.

File: Userland/Libraries/LibGfx/Painter.cpp
Comment (AtkinsSJ) [inline Userland/Libraries/LibGfx/Painter.cpp:1253]: I'm not sure about this. The `do_draw_scaled_bitmap()` this calls doesn't have any handling for `Painter::ScalingMode::None`, so I think it will fall back to some kind of scaling, which may or may not even be valid.

I don't know what's correct here. Maybe `VERIFY()` here that the `src_rect` and `dst_rect` are the same size, and then call `blit_bitmap()` instead.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Sorry it's taken me so long to get back to this! I am easily distracted. :sweat_smile:

When fixing code in response to a review, please modify the original commits instead of adding new ones. (So, rather than adding the bool parameter in one commit, then removing it again in a later commit, just modify the first commit to not add it in the first place.) [This video](https://www.youtube.com/watch?v=ElRzTuYln0M) is helpful if you are not familiar with how to do this. You can also ask in one of the Discord channels if you get stuck.

**Comment by AtkinsSJ** [comment]
Oh also, please click the "Resolve conversation" button on any review comments that you have dealt with. It makes it easier for maintainers to spot that they should review the PR again. :^)

File: Userland/Libraries/LibGfx/Painter.cpp
Comment (jpgleeson) [reply Userland/Libraries/LibGfx/Painter.cpp:1253]: Added some handling to `draw_scaled_bitmap` to handle these situations. What it does is create a "clipped" rect of the source to the destination dimensions, and blits that to the destination location. Obviously there's no current code that calls this in Serenity, but a minimum sample would be something like the diff below. Then you can run FlappyBug and verify that the cloud is clipped to half it's height. You can change the `0.5f` value in `Game.h` to for example `2.5f` and see that we don't scale the image to that dimension, we just show what we have in the source at it's original scale.

image


```
diff --git a/Userland/Games/FlappyBug/Game.cpp b/Userland/Games/FlappyBug/Game.cpp
index e8fc4946bd..5ed660dd8d 100644
--- a/Userland/Games/FlappyBug/Game.cpp
+++ b/Userland/Games/FlappyBug/Game.cpp
@@ -63,7 +63,7 @@ void Game::paint_event(GUI::PaintEvent& event)
 
     painter.draw_tiled_bitmap(frame_inner_rect(), *m_background_bitmap);
 
-    painter.draw_scaled_bitmap(m_cloud.rect(), *m_cloud.bitmap(), m_cloud.bitmap()->rect(), 0.2f);
+    painter.draw_scaled_bitmap(m_cloud.rect(), *m_cloud.bitmap(), m_cloud.bitmap()->rect(), 0.2f, Gfx::Painter::ScalingMode::None);
 
     painter.fill_rect(enclosing_int_rect(m_obstacle.top_rect()), m_obstacle.color);
     painter.fill_rect(enclosing_int_rect(m_obstacle.bottom_rect()), m_obstacle.color);
diff --git a/Userland/Games/FlappyBug/Game.h b/Userland/Games/FlappyBug/Game.h
index 74137707b9..95684fd9d1 100644
--- a/Userland/Games/FlappyBug/Game.h
+++ b/Userland/Games/FlappyBug/Game.h
@@ -160,7 +160,7 @@ public:
 
         Gfx::IntRect rect() const
         {
-            return { (int)x - bitmap()->width(), (int)y - bitmap()->height(), bitmap()->width(), bitmap()->height() };
+            return { (int)x - bitmap()->width(), (int)y - bitmap()->height(), bitmap()->width(), (int)(bitmap()->height()  * 0.5f)};
         }
     };
 

```

**Review by AtkinsSJ — APPROVED** [review]
LGTM!

**Pull request merged**
